kernel/src/sched.rs

```diff
@@ -50,11 +108,28 @@ pub struct Kernel {
     grants_finalized: Cell<bool>,
 }
 
+/// Enum used to inform scheduler why a process stopped
+/// executing (aka why `do_process()` returned).
+#[derive(PartialEq, Eq)]
+pub enum StoppedExecutingReason {
+    /// The process returned because it is no longer ready to run
+    NoWorkLeft,
+
+    /// The process returned because its timeslice expired
+    TimesliceExpired,
+
+    /// The process returned because it was preempted by kernel
+    /// work that became ready (most likely because an interrupt fired
+    /// and the kernel thread needs to execute the bottom half of the
+    /// interrupt)
+    KernelPreemption,
+}
```

Comment:
What about it crashed or exited?

Reply:
Those are currently grouped in as part of `NoWorkLeft`, as none of the schedulers I have implemented do anything different for crashed/exited than they do for yielded. But it makes sense that some schedulers might want to handle those cases differently (I could imagine optimization of scheduler internal data structures by removing nodes that won't be scheduled, or a scheduler punishing crashed processes even after they restart), so I will add them in.

Reply:
Though, I am not sure adding "Exited" makes sense until Tock 2.0 when we have an Exited syscall

Reply:
Added `StoppedFaulted` and `Stopped`. I specified `StoppedFaulted` bc restarting faults will not cause do_process to exit